README.md
[![npm](https://img.shields.io/npm/v/swagger-ui-cli.svg)](https://www.npmjs.com/package/swagger-ui-cli)

# swagger-ui-cli

A standalone CLI application, serving [Swagger UIs](https://swagger.io/tools/swagger-ui/) via a HTTP server.

## Install

You can install it globally:

```bash
npm install -g swagger-ui-cli
```

Or for your project, from where your `package.json` file is stored:

```bash
npm install --save-dev swagger-ui-cli
```

## Usage

```
$ swagger-ui --help

  A standalone CLI application, serving Swagger UIs via a HTTP server.

  Usage
    $ swagger-ui [options] <file>

  Options
    --allow-scripts    Allow the execution of scripts. Default: (false)
    --do-not-open      Do not open local URL after server has been started. Default: (false)
    --no-json          Do not provide JSON data as download. Default: (false)
    --no-toml          Do not provide TOML data as download. Default: (false)
    --no-yaml          Do not provide YAML data as download. Default: (false)
    --port, -p         The custom TCP port. Default: 8080

  <file> The source document as local file path or URL. Supports JSON, YAML and TOML.

  Examples
    Starts a new server instance on port 8080 for a local file
    $ swagger-ui swaggerFile.yaml

    Run a Node.js script (also from a remote host), which builds the document
    $ swagger-ui buildDoc.js --allow-scripts

    Using port 8181 and load document from HTTP server without providing TOML as download
    $ swagger-ui https://raw.githubusercontent.com/apache/superset/1.4.0rc2/docs/src/resources/openapi.json --no-toml --port=8181

    Do not open browser, after server has been started and use "foo" as username and "bar" as password
    $ swagger-ui https://example.com/my-api.toml --do-not-open --username=foo --password=bar
```

### Scripts

If you have a complex logic to build an [OpenAPI](https://www.openapis.org/) document, maybe it is separated into multiply sources and parts, you can execute JavaScript code, which runs in the same Node.js environment as the application.

In that case, you have to start the application with `--allow-scripts` flag.

Example:

```javascript
// use any Node.js you want
const fs = require("fs");
// you are also able to access 3rd party modules
// if a 'node_modules' folder is available
const axios = require("axios");
// make use of local Node modules
// which exports functions that loads
// parts of the document, e.g.
const myModule = require("/path/to/my/module.js");

const info = await fs.promises.readFile(
  "/path/to/apiDocumentInfo.json",
  "utf8"
);

// maybe load data from remote sources
const paths = (await axios.get("https://strapi.example.com/paths")).data;
const components = await myModule.loadComponents();

// put all parts together ...
const doc = {
  openapi: "3.0.0",

  info,

  servers: [
    {
      url: "http://petstore.swagger.io/api",
    },
  ],

  paths,
  components,
};

// ... and return the document
return doc;
```

## Download

You can download documents via browser or HTTP client directly. Examples:

- http://localhost:8080/json ([JSON](https://en.wikipedia.org/wiki/JSON))
- http://localhost:8080/toml ([TOML](https://en.wikipedia.org/wiki/TOML))
- http://localhost:8080/yaml ([YAML](https://en.wikipedia.org/wiki/YAML))

## Contributors

- [malagege](https://github.com/malagege)
- [Geno Roupsky](https://github.com/groupsky)

## License

[GPL 3.0](./LICENSE)


CHANGELOG.md
# Change Log (swagger-ui-cli)

## 0.15.0

- **BREAKING CHANGE**: module requires at least [Node 16+](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V16.md) now
- `npm update`s

## 0.14.0

- [Fix Swagger UI OAuth2 Redirect URL with parameters return 404 Response.](https://github.com/egomobile/swagger-ui-cli/pull/3) by [malagege](https://github.com/malagege)
- `npm update`s

## 0.12.2

- add optional `no-json`, `no-toml` and `no-yaml` CLI flags, s. [issue #2](https://github.com/egomobile/swagger-ui-cli/issues/2)
- replace [mime-types](https://www.npmjs.com/package/mime-types) with [mrmime](https://www.npmjs.com/package/mrmime)
- apply new [eslint settings](https://github.com/egomobile/eslint-config-ego)
- `npm update`s
- code cleanups and improvements

## 0.11.1

- application requires at least [Node.js 14](https://nodejs.org/es/blog/release/v14.0.0/) now
- `npm update`s
- (bug-)fixes

## 0.10.0

- application requires at least [Node.js 12](https://nodejs.org/es/blog/release/v12.0.0/) now
- (bug-)fixes
- made all Node dependencies explicit

## 0.9.1

- CLI requires a least [Node.js 12+](https://nodejs.org/en/blog/release/v12.0.0/) now
- change owner to [e.GO Mobile](https://e-go-mobile.com/)
- `npm update`

## 0.8.0

- can execute JavaScript code now, to build a Swagger document

## 0.7.0

- [Bump axios to v0.21.1](https://github.com/egodigital/swagger-ui-cli/pull/1) by [Geno Roupsky](https://github.com/groupsky)

## 0.6.0

- fix README and CLI help screen
- update npm modules:
  - `ora^5.1.0`,
  - `swagger-ui-dist^3.33.0`

## 0.5.3

- bugfixes
- refactoring
  - code cleanups
  - improvements

## 0.4.5

- initial release


src/cli.ts
/**
 * swagger-ui-cli - Standalone Swagger UI server
 * Copyright (C) 2021  Next.e.GO Mobile SE, Aachen, Germany
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import meow from "meow";
import { ExitCode } from "./contracts";

const cli = meow(`
Usage
  $ swagger-ui [options] <file>

Options
  --allow-scripts    Allow the execution of scripts. Default: (false)
  --do-not-open      Do not open local URL after server has been started. Default: (false)
  --no-json          Do not provide JSON data as download. Default: (false)
  --no-toml          Do not provide TOML data as download. Default: (false)
  --no-yaml          Do not provide YAML data as download. Default: (false)
  --password         The password for authentication.
  --port, -p         The custom TCP port. Default: 8080
  --username         The username for authentication.

<file> The source document as local file path or URL. Supports JSON, YAML and TOML.

Examples
  Starts a new server instance on port 8080 for a local file
  $ swagger-ui swaggerFile.yaml

  Run a Node.js script (also from a remote host), which builds the document
  $ swagger-ui buildDoc.js --allow-scripts

  Using port 8181 and load document from HTTP server without providing TOML as download
  $ swagger-ui https://raw.githubusercontent.com/apache/superset/1.4.0rc2/docs/src/resources/openapi.json --no-toml --port=8181

  Do not open browser, after server has been started and use "foo" as username and "bar" as password
  $ swagger-ui https://example.com/my-api.toml --do-not-open --username=foo --password=bar
`, {
    "flags": {
        "allow-scripts": {
            "type": "boolean",
            "default": false,
            "isRequired": false
        },
        "do-not-open": {
            "type": "boolean",
            "default": false,
            "isRequired": false
        },
        "json": {
            "type": "boolean",
            "default": true,
            "isRequired": false
        },
        "password": {
            "type": "string",
            "isRequired": false
        },
        "port": {
            "type": "number",
            "alias": "p",
            "default": 8080,
            "isRequired": false
        },
        "toml": {
            "type": "boolean",
            "default": true,
            "isRequired": false
        },
        "yaml": {
            "type": "boolean",
            "default": true,
            "isRequired": false
        },
        "username": {
            "type": "string",
            "isRequired": false
        }
    }
});

/**
 * Exists with a code and a warning message.
 *
 * @param {ExitCode} code The code.
 * @param {any} msg The message.
 */
export function exitWith(code: ExitCode, msg: any) {
    console.warn("🛑", msg);

    console.log("");
    cli.showHelp(code);
}

export default cli;